crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs

```diff
@@ -66,57 +88,214 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
     }
 }
 
-#[derive(Debug, Clone, Copy, Eq, PartialEq)]
-enum Redundancy {
-    FloatComplex,
-    IntComplex,
-    IntFloat,
-}
-
-fn check_annotation(checker: &mut Checker, annotation: &Expr) {
-    let mut has_float = false;
-    let mut has_complex = false;
-    let mut has_int = false;
+fn check_annotation<'a>(checker: &mut Checker, annotation: &'a Expr) {
+    let mut numeric_flags = NumericFlags::empty();
 
     let mut find_numeric_type = |expr: &Expr, _parent: &Expr| {
         let Some(builtin_type) = checker.semantic().resolve_builtin_symbol(expr) else {
             return;
         };
 
-        match builtin_type {
-            "int" => has_int = true,
-            "float" => has_float = true,
-            "complex" => has_complex = true,
-            _ => {}
-        }
+        numeric_flags.seen_builtin_type(builtin_type);
     };
 
+    // Traverse the union, and remember which numeric types are found.
     traverse_union(&mut find_numeric_type, checker.semantic(), annotation);
 
-    if has_complex {
-        if has_float {
-            checker.diagnostics.push(Diagnostic::new(
-                RedundantNumericUnion {
-                    redundancy: Redundancy::FloatComplex,
-                },
-                annotation.range(),
-            ));
+    let Some(redundancy) = Redundancy::from_numeric_flags(numeric_flags) else {
+        return;
+    };
+
+    // Traverse the union a second time to construct the fix.
+    let mut necessary_nodes: Vec<&Expr> = Vec::new();
+
+    let mut union_type = UnionKind::TypingUnion;
+    let mut remove_numeric_type = |expr: &'a Expr, parent: &'a Expr| {
+        let Some(builtin_type) = checker.semantic().resolve_builtin_symbol(expr) else {
+            // Keep type annotations that are not numeric.
+            necessary_nodes.push(expr);
+            return;
+        };
+
+        if matches!(parent, Expr::BinOp(_)) {
+            union_type = UnionKind::PEP604;
         }
 
-        if has_int {
-            checker.diagnostics.push(Diagnostic::new(
-                RedundantNumericUnion {
-                    redundancy: Redundancy::IntComplex,
-                },
-                annotation.range(),
-            ));
+        // `int` is always dropped, since `float` or `complex` must be present.
+        // `float` is only dropped if `complex`` is present.
+        if (builtin_type == "float" && !numeric_flags.contains(NumericFlags::COMPLEX))
+            || (builtin_type != "float" && builtin_type != "int")
```

Comment:
Isn't this the same as?

```suggestion
            || builtin_type == "complex"
```

Nit: It could also be worth to have a `BuiltInTy` enum with `Float`, `Int`, and `Complex` to reduce some of the string matching

Reply:
Here I think the `builtin_type` is not guaranteed to be `int`/`float`/`complex`, in which case we cannot replace this condition

Reply:
Oh, in case it is e.g. `str` or `bool` and I think there are unit tests that exercise this.